### ISSUE TITLE
[TASK] PSR-7 for Ajax requests/responses

### ISSUE THREAD

**Issue by thomaszbz** [mblaschke/TYPO3-metaseo#113]
TYPO3 CMS 7.4 claims to be PSR-7 compliant:

"TYPO3 7.4 is the first Open Source Content Management System which uses the newly introduced PSR-7 standard, which allows for a proper Request/Response logic known from other standalone PHP frameworks. All entry points have been unified to use a clean Application class, so that no PHP code is called in the global scope. The TYPO3 Backend now only uses typo3/index.php for calling anything (AJAX Requests, Backend Modules, etc) with a flexible Request Handling API."

We could see which components of TYPO3 CMS we could reuse. Plus, we might learn a lot about the new clean code of TYPO3 7.4. Hopefully.

I already suggested to use of OOP for request and response objects in #92.

However, we must check if all the new jazz introduced in TYPO3 7.4 still works with TYPO3 6.2. Basically this could mean that this refactoring step cannot be done until EOL of 6.2, at least not the way they did it in 7.4. As said, we need to investigate on that question.

References:
http://www.php-fig.org/psr/psr-7/
http://typo3.org/news/article/announcing-typo3-74-nothing-is-impossible/

**Comment by thomaszbz** [comment]
needs at least one additional dependency which TYPO3 6.2 does not have, but TYPO3 7 does:

    "require": {
        "psr/http-message": "~1.0"
    },

We could add this to composer.json to get it work in 6.2 but I'm not sure if everybody already uses composer.

**Comment by thomaszbz** [comment]
Composer documentation: https://wiki.typo3.org/Composer

Best practise is to use multiple branches for 6.2 and 7 in terms of composer. At least not to use enhanced features for 6.2. Best practise might even be to not use enhanced features of coposer at all (including for TYPO3 7) until 6.2's EOL at least.

Full-featured composer.json could look like this: https://github.com/helhum/typo3_console/blob/master/composer.json. Note there is al branch for TYPO3 6.2: https://github.com/helhum/typo3_console/blob/TYPO3_7-3/composer.json. And there is no composer.lock for both of them. In case we need to provide the clases in our release packages I'd rather see them locked so that we can track changes via SCM. A composer.json without a composer.lock is always dangerous.

Compatibility in terms of TYPO3 version (7.x) can be set in composer.json. However, this means that a fork needs to exist for the older TYPO3 version (6.2).

Important:

* composer.json dependencies _must_ be in the filesystem before deployment and they _must_ be shipped in the release package.
* If the core has the same dependency, the core will autoload dependencies first.
* If multiple extensions have the same dependency then it is unclear which one will be first. But the first one will succeed (this potentially renders bugfixing a nightmare)
* TYPO3 6.2 has no composer support at all which basically means that we have to require_once an autoload.php generated by composer. Maybe we have to test first if our classes are autoloaded already
* TYPO3 7.4 at least has psr/http-message.
* Versions between >6.2 and <7.4 do not have to be considered by definition (their risk, their fun).
* Currently, all the core components including dependencies are shipped with the official built package. Composer install only needs to be done in case the TYPO3.CMS repository is checked out via git.
*TYPO3 7 has some mismatch in terms of composer vs. Extension Manager. Currently, Extension Manager is used in TYPO3 7 for dependencies of extensions. This might change in a way that Extension Manager triggers composer's dependency resolution (or at least Extension Manager uses its own implementation to resolve dependencies in composer.json files (composer.lock does not make sense here in case multiple extensions lock the same dependency by commit-id: this somehow can never be resolved.) Basically that means that as soon as Extension manager tries to resolve versions that somehow fits to all reqirements defined in multiple composer.json files, then the chosen version does not necessarily apply to one of the versions provided by the packages which basically means that packages should not provide the classes at all which basically means that if we provide them now we might possible have to remove them later when we use composer for such stuff.

Another option would be to manipulate namespaces of third party code and copy it into our repository. That would be at least the only option I know of which lets us reproduce bugs only against an exact version of the dependency (which we know) and the TYPO3 core (who's dependency versions we also know). In the end this could even be the best option as long as there's a bunch of breaking changes on the road but not backed with a sustainable concept.

**Comment by thomaszbz** [comment]
Related: #147 and #129

**Comment by thomaszbz** [comment]
TYPO3 CMS 6.2 comes along with an ``\TYPO3\CMS\Core\Http\AjaxRequestHandler`` which we can use like this:

    /**
     * Execute ajax call
     */
    public function main($params = array(), \TYPO3\CMS\Core\Http\AjaxRequestHandler &$ajaxObj = NULL)

**Comment by thomaszbz** [comment]
The parameter ``$params`` is an empty array until 7.4.0, c048cede
* https://review.typo3.org/#/c/41145/
* https://forge.typo3.org/projects/typo3cms-core/repository/revisions/c048cede7a402275e501d7a86cb560995f29360b/diff/typo3/sysext/backend/Classes/Http/AjaxRequestHandler.php
* https://forge.typo3.org/issues/68186

Starting with 7.4.0 we can retrieve parameters via ServerRequestInterface
    
    /**
     * @param $params
     *
     * @return \Psr\Http\Message\ServerRequestInterface
     */
    protected function getRequest($params)
    {
        return $params['request'];
    }

I see no way to introduce a ServerRequestInterface in our own namespace and specific for MetaSEO's AJAX calls only. Plus, do we get an empty array here until 7.4.0. Which is why we can't use this until we drop support for TYPO3 CMS versions 6.2.x - 7.3.x

**Comment by thomaszbz** [comment]
As of TYPO3 CMS 7.x, there is a ``TYPO3\CMS\Backend\Http\AjaxRequestHandler``, but it is not injected into ``&$ajaxObj``. Instead ``TYPO3\CMS\Backend\Http\AjaxRequestHandler`` is responsible for injecting ``\TYPO3\CMS\Core\Http\AjaxRequestHandler``, so that nothing changes in that respect.

However, in ``\TYPO3\CMS\Core\Http\AjaxRequestHandler`` there is a limitation: Http status code 500 is hard coded here which is why I filed issue https://forge.typo3.org/issues/69388 for that.

I assume this is not being backported (and even if it would, we cannot rely on users being up-to-date). Again, we cannot use this until we drop support for TYPO3 CMS versions 6.2.x - 7.3.x or use TYPO3 version specific branches.

**Comment by thomaszbz** [comment]
As of TYPO3 CMS v7 final, according to core-devs, there will be no sort of Dependency Injection available when we register Ajax handlers with

    Typo3ExtensionManagementUtility::registerAjaxHandler

Basically that means we should either have a setter for injections which we can use for unit tests. Or we could go for extbase and use their implementation of DI.

On the other side, DI might make it into the core in TYPO3 CMS v8, at least this is in discussion.

I tend to not go for extbase and see if we can use DI provided by the core in v8. There's no big advantage when we use extbase just for DI. And then refactor it again for v8's DI.

**Comment by thomaszbz** [comment]
I decided against extbase-DI for the Ajax-controller, because extbase-DI might get deprecated at some time.
We get CSRF and PSR-7 from the core and get request/response injected by the core. The response object allows content to be rendered to json.
Using an init function which is testing for already injected mocked objects, the class remains testable.
As soon as the core supports DI, the init function can be removed.
Self-made CSRF protection now gets obsolete and, consequently, I removed it.
